example/poisson_gmg/poisson_equation.hpp

```diff
@@ -0,0 +1,218 @@
+//========================================================================================
+// (C) (or copyright) 2023. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef EXAMPLE_POISSON_GMG_POISSON_EQUATION_HPP_
+#define EXAMPLE_POISSON_GMG_POISSON_EQUATION_HPP_
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <kokkos_abstraction.hpp>
+#include <parthenon/package.hpp>
+
+#include "poisson_package.hpp"
+
+namespace poisson_package {
+
+// This class implement methods for calculating A.x = y and returning the diagonal of A,
+// where A is the the matrix representing the discretized Poisson equation on the grid.
+// Here we implement the Laplace operator in terms of a flux divergence to (potentially)
+// consistently deal with coarse fine boundaries on the grid. Only the routines Ax and
+// SetDiagonal need to be defined for interfacing this with solvers. The other methods
+// are internal, but can't be marked private or protected because they launch kernels
+// on device.
```

Comment:
It might be useful to specify somewhere (for example, the docs ;) ) exaclty the API expected of an equations type. Could even impose it with a concept.

Reply:
I added a couple of comments in the the solver headers. It could be imposed with a concept, but I think doing that is a little tricky since these are template methods. Also, it is not a really general thing so a concept might be overkill.

Reply:
Works for me